pkg/credentialprovider/apis/registrycredentials/types.go

```diff
@@ -0,0 +1,115 @@
+limitations under the License.
+*/
+
+package registrycredentials
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+type RegistryCredentialConfig struct {
+	metav1.TypeMeta `json:",inline"`
+
+	Providers []RegistryCredentialProvider `json:"providers"`
+}
+
+// RegistryCredentialProvider is used by the kubelet container runtime to match the
+// image property string (from the container spec) with exec-based credential provider
+// plugins that provide container registry credentials.
+type RegistryCredentialProvider struct {
+
+	// ImageMatchers is a list of strings used to match against the image property
+	// (sometimes called "registry path") to determine which images to provide
+	// credentials for.  If one of the strings matches the image property, then the
+	// RegistryCredentialProvider will be used by kubelet to provide credentials
+	// for the image pull.
+
+	// The image property of a container supports the same syntax as the docker
+	// command does, including private registries and tags. A registry path is
+	// similar to a URL, but does not contain a protocol specifier (https://).
+	//
+	// Each ImageMatcher string is a pattern which can optionally contain
+	// a port and a path, similar to the image spec.  Globs can be used in the
+	// hostname (but not the port or the path).
+	//
+	// Globs are supported as subdomains (*.k8s.io) or (k8s.*.io), and
+	// top-level-domains (k8s.*).  Matching partial subdomains is also supported
+	// (app*.k8s.io).  Each glob can only match a single subdomain segment, so
+	// *.io does not match *.k8s.io.
+	//
+	// The image property matches when it has the same number of parts as the
+	// ImageMatcher string, and each part matches.  Additionally the path of
+	// ImageMatcher must be a prefix of the target URL. If the ImageMatcher
+	// contains a port, then the port must match as well.
+	ImageMatchers []string `json:"imageMatchers"`
+
+	// Exec specifies a custom exec-based plugin.  This type is defined in
+	Exec ExecConfig `json:"exec"`
+}
+
+// ExecConfig specifies a command to provide client credentials. The command is exec'd
+// and outputs structured stdout holding credentials.
+//
+// See the client.authentiction.k8s.io API group for specifications of the exact input
+// and output format
+type ExecConfig struct {
+	// Command to execute.
+	Command string `json:"command"`
+	// Arguments to pass to the command when executing it.
+	// +optional
+	Args []string `json:"args,omitempty"`
+	// Env defines additional environment variables to expose to the process. These
+	// are unioned with the host's environment, as well as variables client-go uses
+	// to pass argument to the plugin.
+	// +optional
+	Env []ExecEnvVar `json:"env,omitempty"`
+
+	// Preferred input version of the ExecInfo. The returned ExecCredentials MUST use
+	// the same encoding version as the input.
+	APIVersion string `json:"apiVersion"`
```

Comment:
Wondering if this should be implied by the APIVersion of  `RegistryCredentialProvider`

Reply:
Yeah you're right, I don't think there is a need for an explicit API version check outside of what the scheme in the kubelet is created with.  I.e. if the Kubelet's decoder can decode it then its a recognized version.  Will remove.